include/deal.II/numerics/history.h

```diff
@@ -0,0 +1,248 @@
+ * memory-demanding objects (i.e. linear algebra vectors), the removal of an
+ * element does not free the memory. Instead the unique pointer to the element
+ * is being kept in a separate cache so that subsequent addition does not
+ * require re-allocation of memory.
+ *
+ * The primary usage of this class is in solvers to store a history of vectors.
+ * That is, if at the iteration $k$ we store $m$ vectors from previous
+ * iterations
+ * $\{k-1,k-2,...,k-m\}$, then addition of the new element will make
+ * the object contain elements from iterations $\{k,k-1,k-2,...,k-m+1\}$.
+ *
+ * @author Denis Davydov, 2018
+ */
+template <typename T>
+class FiniteSizeHistory
+{
+public:
+  static_assert(std::is_default_constructible<T>::value,
+                "Class T requires default-constructible elements");
+
+  /**
+   * Constructor.
+   *
+   * @parameter max_elements maximum number of elements to be stored in the
+   * history.
+   */
+  FiniteSizeHistory(const std::size_t max_elements = 0);
+
+  /**
+   * Add new object by copying.
+   * If the maximum number of elements is reached, the oldest element is
+   * removed.
+   */
+  void
+  add(const T &element);
+
+  /**
+   * Remove an element with index @p index,
+   * counting from the last added element.
+   */
+  void
+  remove(const std::size_t index);
+
+  /**
+   * Read/write access to an element with index @p index,
+   * counting from the last added element.
+   */
+  T &operator[](const std::size_t index);
+
+  /**
+   * Read access to an element with index @p index,
+   * counting from the last added element.
+   */
+  const T &operator[](const std::size_t index) const;
+
+  /**
+   * Return the current size of the history.
+   */
+  std::size_t
+  size() const;
+
+  /**
+   * Return the maximum allowed size of the history.
+   */
+  std::size_t
+  max_size() const;
+
+  /**
+   * Clear the contents, including the cache.
+   */
+  void
+  clear();
+
+private:
+  /**
+   * Maximum number of elements to be stored.
+   */
+  std::size_t max_n_elements;
+
+  /**
+   * A deque which stores the data.
+   */
+  std::deque<std::unique_ptr<T>> data;
+
+  /**
+   * A deque to cache data, in particular for the case when
+   * removal is followed by addition.
+   */
+  std::deque<std::unique_ptr<T>> cache;
+};
+
+
+
+// -------------------  inline and template functions ----------------
+#ifndef DOXYGEN
+
+
+
+template <typename T>
+FiniteSizeHistory<T>::FiniteSizeHistory(const std::size_t max_elements)
+  : max_n_elements(max_elements)
+{}
+
+
+
+template <typename T>
+void
+FiniteSizeHistory<T>::remove(const std::size_t ind)
+{
+  Assert(ind < data.size(), ExcIndexRange(ind, 0, data.size()));
+  auto it = std::move(data[ind]);
+  data.erase(data.begin() + ind);
+
+  cache.push_back(std::move(it));
+
+  // whatever we do, we shall not store more than the maximum number of
+  // elements
+  Assert(data.size() + cache.size() <= max_n_elements, ExcInternalError());
```

Comment:
But how do you ensure this? Above you simply `push_back` onto the cache. If there were already several elements in the cache, then this is now going to fail, isn't it?

Reply:
no, it won't. Since I moved the element from `data` to `cache`, the total size should be unchanged.

The constraint on the maximum number is also being kept in `add()`, at least it should and I am not aware of cases where it would fail.